File: Bio_Informatics_CPP/Task-10/PartialDigestProblem.cpp
#include<iostream>
#include<string>
#include<cstring>
#include<map>
#include<vector>
#include<utility>
#include<random>
#include<set>
#include<algorithm>
#include<cmath>

using namespace std;

typedef map<string, bool>mpsb;
typedef vector<string>vs;
typedef vector<char>vc;
typedef vector<int>vii;
typedef set<int>sii;
typedef map<int, int>mpii;

const int inf = 1e5;

vii L , X;
sii visited;
mpii mp , current;

int formula(int a , int b , int c)
{
    c = -c;
    int D = (b * b) - (4 * a * (c));
    int n = (-b + sqrt(D)) / (2 * a);
    return n;
}

void getFrequency()
{
    for(int i=0; i<L.size(); i++)
    {
        mp[L[i]]++;
    }
}

void PDP(int size)
{
    int i , j , k;
    int n = formula(1, -1, 2*size) , takeminimum = inf;
    //cout << n << "\n";
    for(i=1; i<L.size(); i++)
    {
        int element = L[i];
        takeminimum = inf;

        if(n == X.size())
        {
            break;
        }

        else
        {
            for(j=0; j<X.size(); j++)
            {
                int point = abs(element - X[j]);

                //cout << "element = " << element << " X = " << X[j] << " " << "point = " << point << "\n";

                if(mp[point] > 0)
                {
                    int cnt = 0;
                    current.clear();

                    for(k=0; k<X.size(); k++)
                    {
                        int diff2 = abs(point - X[k]);

                        //cout << "diff2 = " << diff2 << "\n";

                        if(mp[diff2] > 0)
                        {
                            current[diff2]+=1;

                            //cout << "diff2 frequency = " << current[diff2] << "\n";

                            if(mp[diff2] >= current[diff2])
                            {
                                cnt+=1;
                                //cout << "cnt = " << cnt << "\n";
                            }

                            else
                            {
                                continue;
                            }
                        }

                        else
                        {
                            continue;
                        }
                    }

                    if(cnt == X.size())
                    {
                        takeminimum = min(takeminimum , point);
                        //cout << "takeminimum = " << takeminimum << "\n";
                    }
                }

                else
                {
                    continue;
                }
            }

            if(takeminimum < inf)
            {
                X.push_back(takeminimum);
                mp[takeminimum]--;
                //cout << "final minimum = " << takeminimum << "\n";
            }
        }
    }
}

bool checkInX()
{
    for(int i=0; i<X.size(); i++)
    {
        for(int j=i+1; j<X.size(); j++)
        {
            int diff = abs(X[i] - X[j]);

            mp[diff]--;
        }
    }

    bool fl=true;

    for(int i=0; i<X.size(); i++)
    {
        if(mp[X[i]] > 0)
        {
            fl=false;
            break;
        }
    }

    if(fl) return true;

    return false;
}

int main()
{
    cout << ">>>> Start terminal <<<<\n";
    cout << "\nEnter Size(0 to end): ";
    int size;
    cin >> size;
    while(size)
    {
        visited.clear();
        L.clear();
        mp.clear();
        current.clear();
        X.clear();

        cout << "\nEnter " << size << " elements: ";

        while(size--)
        {
            int x;
            cin >> x;
            L.push_back(x);
        }

        sort(L.rbegin(), L.rend());
        //for(int i=0; i<L.size(); i++) cout << L[i] << "; ";

        getFrequency();

        X.push_back(0);
        X.push_back(L[0]);

        PDP(L.size());
        //for(int i=0; i<X.size(); i++) cout << X[i] << "; ";
        bool prove = checkInX();

        if(!prove) cout << "\nNo Solution\n";

        else
        {
            cout << "\nPartial Digest Points are: ";

            sort(X.begin() , X.end());

            for(int i=0; i<X.size(); i++)
            {
                cout << X[i] << " ";
            }

            cout << "\n";
        }

        cout << "\nEnter Size(0 to end): ";
        cin >> size;
    }

    cout << "\n<<<< End terminal >>>>\n";
    return 0;
}

File: BioInfoPython/task-10/Partial Digest Problem.py
# -*- coding: utf-8 -*-
"""
Created on 01-Aug-17 at 3:35 AM

@author: Pranta Sarker

Task: Partial Digest Problem

python-3.6.2

"""         
from math import *;

currfq = {};
count = {};
L = [];

def formula(a , b , c):
    D = b**2 - (4 * a * c);
    n = (-b + sqrt(D)) / (2 * a);
    return int(n);

def init(len):
    global currfq;
    for i in range(0, len+2):
        currfq[i] = 0;
    #print(currfq , end='; ');

def PDP(size):

    global L , count , currfq, X , visited;

    n = formula(1 , -1, -(2 * size));
    # print('n = ' , n);
    # print('X = ' , X);
    # print('coount = ' ,  count);
    # print('L = ' , L);

    for i in range(1 , len(L)+1):

        element = L[i];

        takeminimum = 1e5;

        if(len(X) == n):
            return X;

        else:

            for item1 in X:
                point = abs(element - item1);

                if(point in visited):
                    #print('val1 = ', item1 , ', element = ' , element, ' , point = ', point);
                    cnt = 0;
                    init(L[0]);

                    for item2 in X:

                        diff2 = abs(item2 - point);

                        if(diff2 in visited):

                            #print('point = ' , point , ', item2 = ' , item2, ', diff2 = ', diff2);
                            currfq[diff2] = currfq[diff2] + 1;

                            if(count[diff2] >= currfq[diff2]):
                                cnt+=1;
                                #print('cnt = ' , cnt);
                            else:
                                continue;
                        else:
                            #print('point = ', point, ', item2 = ', item2, ', diff2 = ', diff2 , ' left');
                            continue;

                    if(cnt == len(X)):
                        takeminimum = min(takeminimum , point);
                        #print("cnt and len(X) are equal");
                        #print('temporary minimum = ' , takeminimum);
                else:
                    continue;
            if(takeminimum < 1e5):
                X.append(takeminimum);
                count[takeminimum] -= 1;
                #print('final minimum = ', takeminimum);


def Prove(X):
    global L , visited;

    for i in range(len(X)):
        for j in range(i+1 , len(X)):
            diff = abs(X[i] - X[j]);
            #print('diff = ' , diff);
            count[diff]-=1;

    fl = True;

    for element in visited:
        if(count[element] > 0):
            fl=False;
            break;
    if(fl == True):
        print("Partial Digest Points are: " , X);
    else:
        print("No Solution");



size = int(input('Enter the Size: '));

L = [int(n) for n in input().split()];

#L = list(map(int , input().split()));

L = L[:size];
#print(L);

L.sort(reverse=True);
# print(L);
visited = set(L);
for item in visited:
    count[item] = L.count(item);
# print(getUnique);
# print(L);
# print(count);

X = [0, L[0]];

PDP(size);
X.sort();
# print('X has = ' , X);
Prove(X);